File: e2e-tests/src/test/java/ru/racelog3/e2e_tests/BaseTest.java
package ru.racelog3.e2e_tests;

import static org.openqa.selenium.support.ui.ExpectedConditions.elementToBeClickable;
import static org.openqa.selenium.support.ui.ExpectedConditions.presenceOfElementLocated;
import static org.openqa.selenium.support.ui.ExpectedConditions.presenceOfNestedElementLocatedBy;
import static org.openqa.selenium.support.ui.ExpectedConditions.visibilityOf;

import java.io.File;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.List;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.ie.InternetExplorerDriver;
import org.openqa.selenium.support.ui.WebDriverWait;

public abstract class BaseTest {

	enum Browser {
		CHROME, FIREFOX, IE
	}

	private final static int EXCEPTION_TIMEOUT = 5; // in seconds

	protected final static String HEADER_ENTER_BUTTON = "headerEnterButtonID";
	protected final static String HEADER_ACCOUNT_BUTTON = "headerAccountButtonID";
	protected final static String HEADER_MENU_BUTTON = "headerMenuButtonID";
	protected final static String LIST_ITEM_HOME_BUTTON = "listItemHomeButtonID";
	protected final static String LIST_ITEM_ENTER_BUTTON = "listItemEnterButtonID";
	protected final static String LIST_ITEM_ACCOUNT_BUTTON = "listItemAccountButtonID";
	protected final static String AUTH_MODE_LABEL = "authModeLabelID";
	protected final static String AUTH_NAME = "authNameID";
	protected final static String AUTH_PASSWORD = "authPasswordID";
	protected final static String AUTH_EMAIL = "authEmailID";
	protected final static String AUTH_SUBMIT_BUTTON = "authSubmitButtonID";
	protected final static String AUTH_CHANGE_MODE_LINK = "authChangeModeLinkID";
	protected final static String PROFILE_NAME = "profileNameID";
	protected final static String PROFILE_EMAIL = "profileEmailID";
	protected final static String PROFILE_ROLE = "profileRoleID";
	protected final static String PROFILE_LOGOUT_BUTTON = "profileLogoutButtonID";
	protected final static String RACER_PROFILE_NAME = "racerProfileName";
	protected final static String RACER_PROFILE_BIKE_NUMBER = "racerProfileBikeNumber";
	protected final static String RACER_PROFILE_ADD_REMOVE_BUTTON = "racerProfileAddRemoveButton";
	protected final static String RACER_PROFILES_LIST_EXPAND_BUTTON = "racerProfilesListExpandButton";
	protected final static String RACER_PROFILES_LIST_SUBMIT_BUTTON = "racerProfilesListSubmitButton";
	protected final static String RACE_ITEM_CARD_NAME = "raceItemCardName";
	protected final static String RACE_ITEM_CARD_DATE = "raceItemCardDate";
	protected final static String RACE_ITEM_CARD_MORE_BUTTON = "raceItemCardMoreButton";
	protected final static String RACE_ITEM_INFO_NAME = "raceItemInfoName";
	protected final static String RACE_ITEM_INFO_DATE_LOCATION = "raceItemInfoDateLocationa";
	protected final static String RACE_ITEM_INFO_DESCR = "raceItemInfoDescr";
	protected final static String RACE_PARTICIPANTS_LIST_TABLE = "raceParticipantsListTable";
	protected final static String RACE_PARTICIPANTS_LIST_EXPAND_BUTTON = "raceParticipantsListExpandButton";
	protected final static String RACE_REGISTRATION_LIST_PROFILE_ITEM = "raceRegistrationListProfileItem";
	protected final static String RACE_REGISTRATION_LIST_HEADER = "raceRegistrationListHeader";
	protected final static String RACE_REGISTRATION_LIST_EXPAND_BUTTON = "raceRegistrationListExpandButton";
	protected final static String RACE_REGISTRATION_LIST_SUBMIT_BUTTON = "raceRegistrationListSubmitButton";
	protected final static String ALERT_HEADER = "alertHeader";
	protected final static String ALERT_CONTENT = "alertContent";

	protected final static String RACE_1_NAME = "Grand Prix of France";
	protected final static String RACE_2_NAME = "Grand Prix of Catalunya";
	protected final static String RACE_3_NAME = "Grand Prix of Germany";
	protected final static String RACE_4_NAME = "Grand Prix of Great Britain";

	private WebDriver webDriver;
	private WebDriverWait wait;

	protected WebDriver getDriver() {
		return webDriver;
	}

	@BeforeEach
	void initSelenium() throws URISyntaxException {
		setProperties();
		Browser browser = resolveBrowser();

		switch (browser) {
		case CHROME:
			webDriver = new ChromeDriver();
			break;
		case FIREFOX:
			webDriver = new FirefoxDriver();
			break;
		case IE:
			webDriver = new InternetExplorerDriver();
			break;
		}

		wait = new WebDriverWait(webDriver, EXCEPTION_TIMEOUT);
	}

	@AfterEach
	void closeSelenium() {
		webDriver.quit();

		webDriver = null;
		wait = null;
	}

	private static Browser resolveBrowser() {
		String browser = System.getProperty("racelog3.browser", "CHROME");
		switch (browser) {
		case "CHROME":
			return Browser.CHROME;
		case "FIREFOX":
			return Browser.FIREFOX;
		case "IE":
			return Browser.IE;
		}
		return Browser.CHROME;
	}

	private void setProperties() throws URISyntaxException {
		URL chromeURL = getClass().getClassLoader().getResource("drivers" + File.separator + "chrome" + File.separator
				+ "78.0.3904.105-win32" + File.separator + "chromedriver.exe");
		System.setProperty("webdriver.chrome.driver", Paths.get(chromeURL.toURI()).toFile().getAbsolutePath());

		URL firefoxURL = getClass().getClassLoader().getResource("drivers" + File.separator + "mozilla" + File.separator
				+ "0.26.0-win64" + File.separator + "geckodriver.exe");
		System.setProperty("webdriver.gecko.driver", Paths.get(firefoxURL.toURI()).toFile().getAbsolutePath());

		URL ieURL = getClass().getClassLoader()
				.getResource("drivers" + File.separator + "ie" + File.separator + "IEDriverServer.exe");
		System.setProperty("webdriver.ie.driver", Paths.get(ieURL.toURI()).toFile().getAbsolutePath());
	}

	@Test
	public void runTestImpl() {
		beforeTest();
		testBody();
		afterTest();
	}

	protected void beforeTest() {
		getDriver().get("localhost:3000");
	}

	protected void afterTest() {
		// TODO
	}

	protected abstract void testBody();

	protected String createID(String fieldId, String postfix) {
		return fieldId + "_" + postfix;
	}

	protected void checkElement(String fieldID) {
		wait.until(presenceOfElementLocated(By.id(fieldID)));
	}

	protected void elementDoesNotExist(String fieldID) {
		List<WebElement> deleteLinks = getDriver().findElements(By.id(fieldID));
		Assertions.assertTrue(deleteLinks.isEmpty());
	}

	protected void nestedElementDoesNotExist(String parentFieldID, String childXPath) {
		WebElement parentElement = wait.until(presenceOfElementLocated(By.id(parentFieldID)));
		WebElement childElement = null;
		try {
			childElement = wait.until(presenceOfNestedElementLocatedBy(parentElement, By.xpath(childXPath)));
		} catch (Exception e) {
		}
		Assertions.assertTrue(childElement == null);
	}

	protected void clickElement(String fieldID, boolean waitClickable) {
		WebElement element = wait.until(presenceOfElementLocated(By.id(fieldID)));
		if (waitClickable) {
			element = wait.until(elementToBeClickable(element));
		}
		Assertions.assertTrue(element.isEnabled());
		sleep();
		element.click();
	}

	protected void clickElement(String fieldID) {
		clickElement(fieldID, true);
	}

	protected void checkText(String fieldID, String value, String assertText) {
		WebElement element = wait.until(presenceOfElementLocated(By.id(fieldID)));
		element = wait.until(visibilityOf(element));
		Assertions.assertEquals(value, element.getText(), assertText);
	}

	protected void checkNestedElementTextByXPath(String parentFieldID, String childXPath, String value,
			String assertText) {
		WebElement parentElement = wait.until(presenceOfElementLocated(By.id(parentFieldID)));
		WebElement childElement = wait.until(presenceOfNestedElementLocatedBy(parentElement, By.xpath(childXPath)));
		childElement = wait.until(visibilityOf(childElement));
		Assertions.assertEquals(value, childElement.getText(), assertText);
	}

	protected void checkEnabled(String fieldID, boolean enabled) {
		WebElement element = wait.until(presenceOfElementLocated(By.id(fieldID)));
		Assertions.assertTrue(enabled ? element.isEnabled() : !element.isEnabled());
	}

	protected void typeText(String fieldID, String value) {
		WebElement element = wait.until(presenceOfElementLocated(By.id(fieldID)));
		element = wait.until(elementToBeClickable(element));
		Assertions.assertTrue(element.isEnabled());
		element.clear();
		element.sendKeys(Keys.CONTROL + "a");
		element.sendKeys(Keys.DELETE);
		element.sendKeys(value);
	}

	protected void checkSelected(String fieldID, boolean selected) {
		WebElement element = wait.until(presenceOfElementLocated(By.id(fieldID)));
		Assertions.assertTrue(selected ? element.isSelected() : !element.isSelected());
	}

	protected void backToHomePage() {
		selectHeaderSliderPanelButton(LIST_ITEM_HOME_BUTTON);
	}

	protected void goToProfilePage() {
		selectHeaderSliderPanelButton(LIST_ITEM_ACCOUNT_BUTTON);
	}

	private void selectHeaderSliderPanelButton(String buttonID) {
		clickElement(HEADER_MENU_BUTTON);
		sleep();
		clickElement(buttonID);
		sleep();
	}

	protected void sleep() {
		try {
			Thread.sleep(100);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}

	protected void step(String descr) {
		System.out.println("    " + descr);
	}

	protected void substep(String descr) {
		System.out.println("        " + descr);
	}

}

File: e2e-tests/src/test/kotlin/ru/racelog3/e2e_tests/BaseTest.kt
package ru.racelog3.e2e_tests

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.openqa.selenium.By
import org.openqa.selenium.Keys
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.ie.InternetExplorerDriver
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.Select
import org.openqa.selenium.support.ui.WebDriverWait
import java.io.File
import java.net.URISyntaxException
import java.nio.file.Paths

private const val EXCEPTION_TIMEOUT = 5 // in seconds
private const val LONG_SLEEP_TIMEOUT: Long = 5000;

const val HEADER_ENTER_BUTTON = "headerEnterButtonID"
const val HEADER_ACCOUNT_BUTTON = "headerAccountButtonID"
const val HEADER_MENU_BUTTON = "headerMenuButtonID"
const val LIST_ITEM_HOME_BUTTON = "listItemHomeButtonID"
const val LIST_ITEM_ENTER_BUTTON = "listItemEnterButtonID"
const val LIST_ITEM_ACCOUNT_BUTTON = "listItemAccountButtonID"
const val AUTH_MODE_LABEL = "authModeLabelID"
const val AUTH_NAME = "authNameID"
const val AUTH_PASSWORD = "authPasswordID"
const val AUTH_EMAIL = "authEmailID"
const val AUTH_SUBMIT_BUTTON = "authSubmitButtonID"
const val AUTH_CHANGE_MODE_LINK = "authChangeModeLinkID"
const val PROFILE_NAME = "profileNameID"
const val PROFILE_EMAIL = "profileEmailID"
const val PROFILE_ROLE = "profileRoleID"
const val PROFILE_LOGOUT_BUTTON = "profileLogoutButtonID"
const val RACER_PROFILE_NAME = "racerProfileName"
const val RACER_PROFILE_BIKE_NUMBER = "racerProfileBikeNumber"
const val RACER_PROFILE_ADD_REMOVE_BUTTON = "racerProfileAddRemoveButton"
const val RACER_PROFILES_LIST_EXPAND_BUTTON = "racerProfilesListExpandButton"
const val RACER_PROFILES_LIST_SUBMIT_BUTTON = "racerProfilesListSubmitButton"
const val RACE_ITEM_CARD_NAME = "raceItemCardName"
const val RACE_ITEM_CARD_DATE = "raceItemCardDate"
const val RACE_ITEM_CARD_MORE_BUTTON = "raceItemCardMoreButton"
const val RACE_ITEM_INFO_NAME = "raceItemInfoName"
const val RACE_ITEM_INFO_DATE_LOCATION = "raceItemInfoDateLocationa"
const val RACE_ITEM_INFO_DESCR = "raceItemInfoDescr"
const val RACE_PARTICIPANTS_LIST_TABLE = "raceParticipantsListTable"
const val RACE_PARTICIPANTS_LIST_EXPAND_BUTTON = "raceParticipantsListExpandButton"
const val RACE_REGISTRATION_LIST_PROFILE_ITEM = "raceRegistrationListProfileItem"
const val RACE_REGISTRATION_LIST_HEADER = "raceRegistrationListHeader"
const val RACE_REGISTRATION_LIST_EXPAND_BUTTON = "raceRegistrationListExpandButton"
const val RACE_REGISTRATION_LIST_SUBMIT_BUTTON = "raceRegistrationListSubmitButton"
const val RACE_RESULTS_EXPAND_BUTTON = "raceResultsExpandButton";
const val RACE_RESULTS_HEADER = "raceResultsHeader";
const val RACE_RESULTS_TABLE = "raceResultsTable";
const val ADMIN_RACE_INFO_EXPAND_BUTTON = "adminRaceInfoExpandButton";
const val ADMIN_RACE_INFO_HEADER = "adminRaceInfoHeader";
const val ADMIN_RACE_INFO_NAME_COMBO = "adminRaceInfoNameCombo";
const val ADMIN_RACE_INFO_STATE_COMBO = "adminRaceInfoStateCombo";
const val ADMIN_RACE_INFO_SUBMIT_BUTTON = "adminRaceInfoSubminButton";
const val ALERT_HEADER = "alertHeader"
const val ALERT_CONTENT = "alertContent"

const val RACE_1_NAME = "Grand Prix of France"
const val RACE_2_NAME = "Grand Prix of Catalunya"
const val RACE_3_NAME = "Grand Prix of Germany"
const val RACE_4_NAME = "Grand Prix of Great Britain"

enum class Browser {
    CHROME, FIREFOX, IE
}

private fun resolveBrowser(): Browser {
    when (System.getProperty("racelog3.browser", "CHROME")) {
        "CHROME" -> return Browser.CHROME
        "FIREFOX" -> return Browser.FIREFOX
        "IE" -> return Browser.IE
    }
    return Browser.CHROME
}

abstract class BaseTest {

    private var driver: WebDriver? = null
    private var wait: WebDriverWait? = null

    @BeforeEach
    @Throws(URISyntaxException::class)
    fun initSelenium() {
        setProperties()
        driver = when (resolveBrowser()) {
            Browser.CHROME -> ChromeDriver()
            Browser.FIREFOX -> FirefoxDriver()
            Browser.IE -> InternetExplorerDriver()
        }
        wait = WebDriverWait(driver, EXCEPTION_TIMEOUT.toLong())
    }

    @AfterEach
    fun closeSelenium() {
        driver!!.quit()
        driver = null
        wait = null
    }

    @Throws(URISyntaxException::class)
    private fun setProperties() {
        val chromeURL = javaClass.classLoader.getResource("drivers" + File.separator + "chrome" + File.separator
                + "78.0.3904.105-win32" + File.separator + "chromedriver.exe")
        System.setProperty("webdriver.chrome.driver", Paths.get(chromeURL.toURI()).toFile().absolutePath)
        val firefoxURL = javaClass.classLoader.getResource("drivers" + File.separator + "mozilla" + File.separator
                + "0.26.0-win64" + File.separator + "geckodriver.exe")
        System.setProperty("webdriver.gecko.driver", Paths.get(firefoxURL.toURI()).toFile().absolutePath)
        val ieURL = javaClass.classLoader
                .getResource("drivers" + File.separator + "ie" + File.separator + "IEDriverServer.exe")
        System.setProperty("webdriver.ie.driver", Paths.get(ieURL.toURI()).toFile().absolutePath)
    }

    @Test
    fun runTestImpl() {
        beforeTest()
        testBody()
        afterTest()
    }

    protected fun beforeTest() {
        driver!!["localhost:3000"]
    }

    protected fun afterTest() { // TODO
    }

    protected abstract fun testBody()

    protected fun createID(fieldId: String, postfix: String): String {
        return fieldId + "_" + postfix
    }

    protected fun checkElement(fieldID: String?) {
        wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
    }

    protected fun elementDoesNotExist(fieldID: String?) {
        val deleteLinks = driver!!.findElements(By.id(fieldID))
        Assertions.assertTrue(deleteLinks.isEmpty())
    }

    protected fun nestedElementDoesNotExist(parentFieldID: String?, childXPath: String?) {
        val parentElement = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(parentFieldID)))
        var childElement: WebElement? = null
        try {
            childElement = wait!!.until(ExpectedConditions.presenceOfNestedElementLocatedBy(parentElement, By.xpath(childXPath)))
        } catch (e: Exception) {
        }
        Assertions.assertTrue(childElement == null)
    }

    protected fun nestedElementExists(parentFieldID: String?, childXPath: String?) {
        val parentElement = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(parentFieldID)))
        val childElement = wait!!.until(ExpectedConditions.presenceOfNestedElementLocatedBy(parentElement, By.xpath(childXPath)))
        Assertions.assertTrue(childElement != null)
    }

    protected fun clickElement(fieldID: String?, waitClickable: Boolean = true) {
        var element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        if (waitClickable) {
            element = wait!!.until(ExpectedConditions.elementToBeClickable(element))
        }
        Assertions.assertTrue(element.isEnabled)
        sleep()
        element.click()
    }

    protected fun selectDropdownOption(fieldID: String, value: String) {
        val element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        val dropdown = Select(
                wait!!.until(ExpectedConditions.elementToBeClickable(element))
        )
        dropdown.selectByVisibleText(value)
    }

    protected fun checkText(fieldID: String?, value: String?, assertText: String?) {
        var element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        element = wait!!.until(ExpectedConditions.visibilityOf(element))
        Assertions.assertEquals(value, element.text, assertText)
    }

    protected fun checkNestedElementTextByXPath(parentFieldID: String?, childXPath: String?, value: String?,
                                                assertText: String?) {
        val parentElement = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(parentFieldID)))
        var childElement = wait!!.until(ExpectedConditions.presenceOfNestedElementLocatedBy(parentElement, By.xpath(childXPath)))
        childElement = wait!!.until(ExpectedConditions.visibilityOf(childElement))
        Assertions.assertEquals(value, childElement.text, assertText)
    }

    protected fun checkEnabled(fieldID: String?, enabled: Boolean) {
        val element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        Assertions.assertTrue(if (enabled) element.isEnabled else !element.isEnabled)
    }

    protected fun typeText(fieldID: String?, value: String?) {
        var element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        element = wait!!.until(ExpectedConditions.elementToBeClickable(element))
        Assertions.assertTrue(element.isEnabled)
        element.clear()
        element.sendKeys(Keys.CONTROL.toString() + "a")
        element.sendKeys(Keys.DELETE)
        element.sendKeys(value)
    }

    protected fun checkSelected(fieldID: String?, selected: Boolean) {
        val element = wait!!.until(ExpectedConditions.presenceOfElementLocated(By.id(fieldID)))
        Assertions.assertTrue(if (selected) element.isSelected else !element.isSelected)
    }

    protected fun backToHomePage() {
        selectHeaderSliderPanelButton(LIST_ITEM_HOME_BUTTON)
    }

    protected fun goToProfilePage() {
        selectHeaderSliderPanelButton(LIST_ITEM_ACCOUNT_BUTTON)
    }

    private fun selectHeaderSliderPanelButton(buttonID: String) {
        clickElement(HEADER_MENU_BUTTON)
        sleep()
        clickElement(buttonID)
        sleep()
    }

    protected fun sleep() {
        sleepImpl(100)
    }

    private fun sleepImpl(timeout: Long) {
        try {
            Thread.sleep(timeout)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    protected fun step(descr: String) {
        println("    $descr")
    }

    protected fun substep(descr: String) {
        println("        $descr")
    }
}
